### PR TITLE
Fix commands when used in a workspace

### LINKED ISSUE #4328: yarn outdated not working w/workspaces
**Do you want to request a *feature* or report a *bug*?**

Bug

**What is the current behavior?**

In a project work workspaces, `yarn outdated` claims everything is up to date. (The root project has no dependences). `yarn workspace api outdated` errors with the following:

```
yarn workspace v1.0.0
yarn outdated v1.0.0
error No lockfile in this directory. Run `yarn install` to generate one.
error Command failed.
Exit code: 1
Command: /usr/local/Cellar/node/8.4.0/bin/node
Arguments: /usr/local/Cellar/yarn/1.0.0_1/libexec/lib/cli.js outdated
Directory: /Volumes/git/project/packages/api
Output:

info Visit https://yarnpkg.com/en/docs/cli/workspace for documentation about this command.
```

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**

I expect it to tell me the outdated modules.

**Please mention your node.js, yarn and operating system version.**

yarn 1.0.0

### LINKED ISSUE #4334: "yarn remove" inside workspace fails
**Do you want to request a *feature* or report a *bug*?** Bug

**What is the current behavior?**

Attempting to remove a package from inside a workspace fails with:

```
error No lockfile in this directory. Run `yarn install` to generate one
```

**If the current behavior is a bug, please provide the steps to reproduce.**

```sh
$ cd packages/some-package
$ yarn remove some-dependency
yarn remove v1.0.1
error No lockfile in this directory. Run `yarn install` to generate one
```

**What is the expected behavior?**

The package should be removed and the root lockfile should be updated.

**Please mention your node.js, yarn and operating system version.**

Node 8.4.0
Yarn 1.0.1
Mac OS 10.12

Possibly related to #4328.

### LINKED ISSUE #4348: Adding root dependency removes workspace packages from lockfile
**Do you want to request a *feature* or report a *bug*?** Bug

**What is the current behavior?**

Adding a dev dependency to the root of a workspace adds the new package to the lockfile, but also removes lockfile entries for dependencies of the individual workspaces.

In my project, adding a single dependency to the root trimmed the lockfile length from ~8000 lines to ~4000. The workaround is to use `--no-lockfile` when adding, and then run `yarn install`; this will add just the new dependency to the lockfile, as expected.

**If the current behavior is a bug, please provide the steps to reproduce.**

In a workspace project:

```sh
$ yarn add --dev left-pad # in workspace root
```

This will remove entries from the lockfile that aren't direct dependencies of the root (e.g., are dependencies of the workspaces).

**What is the expected behavior?**

Lockfile is updated with just an entry for the new dependency.

**Workaround**

```sh
$ yarn add --dev --no-lockfile left-pad
$ yarn install
```

This will update `package.json` with the new entry only, and then update the lockfile as part of `yarn install`, which produces the correct results.

**Please mention your node.js, yarn and operating system version.**
Node: 8.4.0
Yarn: 1.0.1
OS: Mac OS 10.12

### PR DESCRIPTION
**Pull request by jgoz**
**Summary**
Fixes #4334 
Fixes #4328 
Fixes #4348
Closes #4584

Previously, `remove`, `outdated` and several related commands would fail when used inside of a workspace package, either directly or via `yarn workspaces <pkg> <cmd> [args]`. These failures were caused by incorrect lockfile resolution, specifically using `config.cwd` rather than `config.lockfileFolder`.

Also fixed:
- When using `remove` inside a workspace, `reinstall` would not clean up extraneous packages.
- When using `outdated`, `upgrade` or `upgrade-interactive` inside of a workspace root or child, the wrong dependencies were often listed and the lockfile was being saved incorrectly after completing an upgrade
- Adding a dev dependency to the workspace root would remove all workspace child dependencies from the lockfile (this PR replaces #4584)

**Test plan**

- Added relevant unit tests
- Existing tests pass
- Manually tested against [repro project](https://github.com/jgoz/yarn-workspace-issue)

### CODE REVIEW

File: src/package-linker.js [hunk 1]
```diff
@@ -246,29 +246,15 @@ export default class PackageLinker {
...
-              possibleExtraneous.add(path.join(filepath, subfile));
-            }
-          } else {
-            possibleExtraneous.add(filepath);
-          }
-        }
-      }
+    // look for extraneous packages in workspaces too
+    if (this.config.lockfileFolder !== this.config.cwd) {
+      await this._readRegistryFolders(this.config.cwd, possibleExtraneous, scopedPaths);
```
Comment (arcanis) [inline src/package-linker.js]: Shouldn't this be a `while` loop that would iterate on each workspace?

Comment (jgoz) [reply src/package-linker.js]: Yes, I think that would make sense. Do you see any value including `cwd` in addition to `lockfileFolder` and all of the workspaces?

Comment (arcanis) [reply src/package-linker.js]: I don't think it's necessary, all the node_modules should be guaranteed to be in these folders.

**Comment by jgoz** [comment]
@arcanis Please note that the scope of this PR increased somewhat as I continued testing commands that could not be executed previously. I have updated the description accordingly.

In order to properly test the now-available commands, my other fix from #4584 had to be included here. @kaylieEB, I would greatly appreciate your insight on this PR because it deviates from the approach you took in #4278.

**Comment by arcanis** [comment]
Closing because of #4630

**Pull request closed without merging**